hphp/runtime/vm/jit/tc-relocate.cpp

```diff
@@ -518,18 +531,30 @@ bool relocateNewTranslation(TransLoc& loc, CodeCache::View cache,
   }
 
   if (asm_count) {
-    x64::adjustForRelocation(rel);
-    x64::adjustMetaDataForRelocation(rel, nullptr, fixups);
-    x64::adjustCodeForRelocation(rel, fixups);
+    adjustForRelocation(rel);
+    adjustMetaDataForRelocation(rel, nullptr, fixups);
+    adjustCodeForRelocation(rel, fixups);
   }
 
   if (debug) {
     auto clearRange = [](TCA start, TCA end) {
       CodeBlock cb;
       cb.init(start, end - start, "Dead code");
-      X64Assembler a {cb};
-      while (cb.available() >= 2) a.ud2();
-      if (cb.available() > 0) a.int3();
+      switch (arch()) {
+        case Arch::PPC64: {
+          ppc64_asm::Assembler a{cb};
+          a.emitTrap();
+          break;
+        }
+        case Arch::X64: {
+          X64Assembler a{cb};
+          a.emitTrap();
+          break;
+        }
+        case Arch::ARM:
+          not_implemented();
+          break;
+      }
```

Comment:
Ah, so it looks like in `Vauto::~Vauto()`, we explicitly do nothing if the unit is completely empty—but in fact, we want the padding behavior.

Try adding the following code after the for loop in that destructor:

```
if (!main().closed()) main() << fallthru{};
if (!cold().closed()) cold() << fallthru{};

switch (arch()) {
  case Arch::X64:
    emitX64(unit(), m_text, m_fixups, nullptr);
    break;
  case Arch::ARM:
    emitARM(unit(), m_text, m_fixups, nullptr);
    break;
  case Arch::PPC64:
    emitPPC64(unit(), m_text, m_fixups, nullptr);
    break;
}
```